File: square-app/js/square-api.js
// Square API Client simplificado
// Soporta dos modos:
// 1. OAuth (App Marketplace) - cada merchant autoriza su cuenta
// 2. Proxy directo (fallback) - usa squareApiCall del proyecto principal

let SQUARE_CONFIG = null;
let squareApiCall = null;
let oauthMode = false; // true si usa OAuth, false si usa proxy directo

// Inicializar con OAuth (App Marketplace)
function initSquareApiWithOAuth(accessToken, locationId) {
    oauthMode = true;
    SQUARE_CONFIG = {
        accessToken: accessToken,
        locationId: locationId,
        apiBaseUrl: 'https://connect.squareup.com'
    };
    console.log('✅ Usando OAuth (App Marketplace)');
    return true;
}

// Inicializar usando el proxy existente (fallback)
function initSquareApi() {
    // Si ya está inicializado con OAuth, no cambiar
    if (oauthMode && SQUARE_CONFIG) {
        return true;
    }
    
    // Intentar cargar desde el script principal si está disponible
    if (typeof window !== 'undefined') {
        if (window.squareApiCall && window.SQUARE_CONFIG) {
            squareApiCall = window.squareApiCall;
            SQUARE_CONFIG = window.SQUARE_CONFIG;
            oauthMode = false;
            console.log('✅ Usando squareApiCall del proyecto principal');
            return true;
        }
    }
    
    // Si no está disponible, cargar dinámicamente
    return loadSquareConfig();
}

// Cargar configuración de Square dinámicamente
async function loadSquareConfig() {
    try {
        // Cargar el script de configuración de Square
        const script = document.createElement('script');
        script.src = '../js/square-config.js';
        document.head.appendChild(script);
        
        return new Promise((resolve) => {
            script.onload = () => {
                if (window.squareApiCall && window.SQUARE_CONFIG) {
                    squareApiCall = window.squareApiCall;
                    SQUARE_CONFIG = window.SQUARE_CONFIG;
                    console.log('✅ Configuración de Square cargada');
                    resolve(true);
                } else {
                    console.error('❌ No se pudo cargar la configuración de Square');
                    resolve(false);
                }
            };
            script.onerror = () => {
                console.error('❌ Error cargando square-config.js');
                resolve(false);
            };
        });
    } catch (error) {
        console.error('Error cargando configuración:', error);
        return false;
    }
}

// Función genérica para llamar a Square API
async function squareApiRequest(endpoint, method = 'GET', body = null) {
    // Si está en modo OAuth, hacer llamada directa
    if (oauthMode && SQUARE_CONFIG && SQUARE_CONFIG.accessToken) {
        const url = `${SQUARE_CONFIG.apiBaseUrl}${endpoint}`;
        const options = {
            method: method,
            headers: {
                'Authorization': `Bearer ${SQUARE_CONFIG.accessToken}`,
                'Content-Type': 'application/json',
                'Square-Version': '2024-01-18'
            }
        };
        
        if (body) {
            options.body = JSON.stringify(body);
        }
        
        try {
            const response = await fetch(url, options);
            
            if (!response.ok) {
                const errorData = await response.json().catch(() => ({}));
                throw new Error(errorData.errors?.[0]?.detail || `Error ${response.status}: ${response.statusText}`);
            }
            
            return await response.json();
        } catch (error) {
            console.error('Error en Square API (OAuth):', error);
            throw error;
        }
    }
    
    // Si no está en modo OAuth, usar proxy
    if (!squareApiCall) {
        const loaded = await initSquareApi();
        if (!loaded) {
            throw new Error('No se pudo cargar la configuración de Square API');
        }
    }
    
    return squareApiCall(endpoint, method, body);
}

// ============ INVENTARIO ============

// Obtener inventario de todos los productos
async function getInventory() {
    try {
        if (!SQUARE_CONFIG) {
            await initSquareApi();
        }
        
        // Primero obtener todos los items
        const catalogResponse = await squareApiRequest('/v2/catalog/search', 'POST', {
            object_types: ['ITEM'],
            limit: 1000
        });
        
        const items = catalogResponse.objects || [];
        const variationIds = [];
        
        // Extraer IDs de variaciones
        items.forEach(item => {
            if (item.item_data?.variations) {
                item.item_data.variations.forEach(variation => {
                    variationIds.push(variation.id);
                });
            }
        });
        
        if (variationIds.length === 0) {
            return [];
        }
        
        // Obtener inventario en lotes
        const inventoryResponse = await squareApiRequest('/v2/inventory/batch-retrieve-counts', 'POST', {
            catalog_object_ids: variationIds,
            location_ids: [SQUARE_CONFIG.locationId]
        });
        
        const inventoryCounts = inventoryResponse.counts || [];
        
        // Combinar items con inventario
        const inventory = [];
        items.forEach(item => {
            if (item.item_data?.variations) {
                item.item_data.variations.forEach(variation => {
                    const inventoryCount = inventoryCounts.find(
                        count => count.catalog_object_id === variation.id
                    );
                    
                    inventory.push({
                        itemId: item.id,
                        variationId: variation.id,
                        name: item.item_data.name,
                        variationName: variation.item_variation_data?.name || 'Default',
                        categoryId: item.item_data.category_id,
                        categoryName: item.item_data.categories?.[0]?.name || 'Sin categoría',
                        price: variation.item_variation_data?.price_money?.amount || 0,
                        quantity: inventoryCount ? parseInt(inventoryCount.quantity) : 0,
                        state: inventoryCount?.state || 'NONE',
                        sku: variation.item_variation_data?.sku || ''
                    });
                });
            }
        });
        
        return inventory;
    } catch (error) {
        console.error('Error obteniendo inventario:', error);
        throw error;
    }
}

// Actualizar inventario
async function updateInventory(variationId, quantity, adjustmentType = 'set') {
    try {
        if (!SQUARE_CONFIG) {
            await initSquareApi();
        }
        
        let finalQuantity = quantity;
        
        if (adjustmentType === 'add' || adjustmentType === 'subtract') {
            // Obtener cantidad actual
            const currentResponse = await squareApiRequest('/v2/inventory/batch-retrieve-counts', 'POST', {
                catalog_object_ids: [variationId],
                location_ids: [SQUARE_CONFIG.locationId]
            });
            
            const currentCount = currentResponse.counts?.[0];
            const currentQuantity = currentCount ? parseInt(currentCount.quantity) : 0;
            
            if (adjustmentType === 'add') {
                finalQuantity = currentQuantity + quantity;
            } else {
                finalQuantity = Math.max(0, currentQuantity - quantity);
            }
        }
        
        // Crear ajuste de inventario
        const response = await squareApiRequest('/v2/inventory/batch-change', 'POST', {
            idempotency_key: `inventory_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            changes: [{
                type: 'ADJUSTMENT',
                adjustment: {
                    catalog_object_id: variationId,
                    catalog_object_type: 'ITEM_VARIATION',
                    from_state: 'NONE',
                    to_state: 'NONE',
                    location_id: SQUARE_CONFIG.locationId,
                    quantity: finalQuantity.toString(),
                    occurred_at: new Date().toISOString()
                }
            }]
        });
        
        return response;
    } catch (error) {
        console.error('Error actualizando inventario:', error);
        throw error;
    }
}

// ============ PEDIDOS ============

// Obtener pedidos
async function getOrders(statusFilter = null) {
    try {
        const query = {
            location_ids: [squareApiClient.locationId],
            query: {
                filter: {}
            },
            limit: 100
        };
        
        if (statusFilter) {
            query.query.filter.state_filter = {
                states: [statusFilter]
            };
        }
        
        const response = await squareApiRequest('/v2/orders/search', 'POST', {
            location_ids: [SQUARE_CONFIG.locationId],
            query: {
                filter: statusFilter ? {
                    state_filter: {
                        states: [statusFilter]
                    }
                } : {}
            },
            limit: 100
        });
        return response.orders || [];
    } catch (error) {
        console.error('Error obteniendo pedidos:', error);
        throw error;
    }
}

// ============ PRODUCTOS ============

// Obtener productos
async function getProducts() {
    try {
        const response = await squareApiRequest('/v2/catalog/search', 'POST', {
            object_types: ['ITEM'],
            limit: 1000
        });
        
        return response.objects || [];
    } catch (error) {
        console.error('Error obteniendo productos:', error);
        throw error;
    }
}

// Crear producto
async function createProduct(productData) {
    try {
        const catalogObject = {
            type: 'ITEM',
            id: `#${Date.now()}`,
            item_data: {
                name: productData.name,
                category_id: productData.categoryId || null,
                variations: [{
                    type: 'ITEM_VARIATION',
                    id: `#${Date.now()}_var`,
                    item_variation_data: {
                        name: 'Default',
                        pricing_type: 'FIXED_PRICING',
                        price_money: {
                            amount: Math.round(productData.price * 100), // Convertir a centavos
                            currency: 'USD'
                        },
                        sku: productData.sku || null,
                        track_inventory: productData.inventory > 0
                    }
                }]
            }
        };
        
        const response = await squareApiRequest('/v2/catalog/object', 'POST', {
            idempotency_key: `product_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            object: catalogObject
        });
        
        // Si hay inventario inicial, agregarlo
        if (productData.inventory > 0 && response.catalog_object?.item_data?.variations?.[0]?.id) {
            await updateInventory(
                response.catalog_object.item_data.variations[0].id,
                productData.inventory,
                'set'
            );
        }
        
        return response;
    } catch (error) {
        console.error('Error creando producto:', error);
        throw error;
    }
}

// Obtener categorías
async function getCategories() {
    try {
        const response = await squareApiRequest('/v2/catalog/search', 'POST', {
            object_types: ['CATEGORY'],
            limit: 1000
        });
        
        return (response.objects || []).map(cat => ({
            id: cat.id,
            name: cat.category_data.name
        }));
    } catch (error) {
        console.error('Error obteniendo categorías:', error);
        throw error;
    }
}

// Exportar funciones
window.squareApi = {
    init: initSquareApi,
    initWithOAuth: initSquareApiWithOAuth, // Para App Marketplace
    getInventory,
    updateInventory,
    getOrders,
    getProducts,
    createProduct,
    getCategories,
    getLocationId: () => SQUARE_CONFIG?.locationId || (oauthMode ? SQUARE_CONFIG?.locationId : null)
};

File: pos_app_flutter/lib/services/square_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class SquareAPI {
  // URL del proxy de Supabase
  static const String _proxyUrl = 'https://fbbvfzeyhhopdwzsooew.supabase.co/functions/v1/square-proxy';
  
  // Configuración de Square (se puede guardar en SharedPreferences)
  static String? _locationId;
  
  static Future<void> init() async {
    final prefs = await SharedPreferences.getInstance();
    _locationId = prefs.getString('square_location_id') ?? 'L94DY3ZD6WS85';
  }
  
  static Future<Map<String, dynamic>> _makeRequest(
    String endpoint,
    String method,
    Map<String, dynamic>? body,
  ) async {
    try {
      final url = Uri.parse('$_proxyUrl$endpoint');
      
      final request = http.Request(method, url);
      request.headers['Content-Type'] = 'application/json';
      
      if (body != null) {
        request.body = jsonEncode(body);
      }
      
      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      } else {
        throw Exception('Error ${response.statusCode}: ${response.body}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }
  
  // Obtener inventario
  static Future<List<Map<String, dynamic>>> getInventory() async {
    await init();
    
    // Obtener productos
    final catalogResponse = await _makeRequest(
      '/v2/catalog/search',
      'POST',
      {
        'object_types': ['ITEM'],
        'limit': 1000,
      },
    );
    
    final items = (catalogResponse['objects'] as List?) ?? [];
    final variationIds = <String>[];
    final itemsMap = <String, Map<String, dynamic>>{};
    
    // Extraer variaciones
    for (var item in items) {
      final itemData = item['item_data'] as Map<String, dynamic>?;
      if (itemData != null && itemData['variations'] != null) {
        final variations = itemData['variations'] as List;
        for (var variation in variations) {
          final variationId = variation['id'] as String;
          variationIds.add(variationId);
          itemsMap[variationId] = {
            'itemId': item['id'],
            'variationId': variationId,
            'name': itemData['name'] ?? 'Sin nombre',
            'variationName': variation['item_variation_data']?['name'] ?? 'Default',
            'categoryName': 'Sin categoría',
            'price': variation['item_variation_data']?['price_money']?['amount'] ?? 0,
            'sku': variation['item_variation_data']?['sku'] ?? '',
          };
        }
      }
    }
    
    if (variationIds.isEmpty) {
      return [];
    }
    
    // Obtener inventario
    final inventoryResponse = await _makeRequest(
      '/v2/inventory/batch-retrieve-counts',
      'POST',
      {
        'catalog_object_ids': variationIds,
        'location_ids': [_locationId],
      },
    );
    
    final counts = (inventoryResponse['counts'] as List?) ?? [];
    
    // Combinar datos
    final inventory = <Map<String, dynamic>>[];
    for (var count in counts) {
      final variationId = count['catalog_object_id'] as String;
      if (itemsMap.containsKey(variationId)) {
        final item = itemsMap[variationId]!;
        item['quantity'] = int.tryParse(count['quantity']?.toString() ?? '0') ?? 0;
        item['state'] = count['state'] ?? 'NONE';
        inventory.add(item);
      }
    }
    
    return inventory;
  }
  
  // Actualizar inventario
  static Future<void> updateInventory(
    String variationId,
    int quantity,
  ) async {
    await init();
    
    await _makeRequest(
      '/v2/inventory/batch-change',
      'POST',
      {
        'idempotency_key': 'inventory_${DateTime.now().millisecondsSinceEpoch}',
        'changes': [
          {
            'type': 'ADJUSTMENT',
            'adjustment': {
              'catalog_object_id': variationId,
              'catalog_object_type': 'ITEM_VARIATION',
              'from_state': 'NONE',
              'to_state': 'NONE',
              'location_id': _locationId,
              'quantity': quantity.toString(),
              'occurred_at': DateTime.now().toIso8601String(),
            },
          },
        ],
      },
    );
  }
  
  // Obtener pedidos
  static Future<List<Map<String, dynamic>>> getOrders() async {
    await init();
    
    final response = await _makeRequest(
      '/v2/orders/search',
      'POST',
      {
        'location_ids': [_locationId],
        'query': {
          'filter': {
            'state_filter': {
              'states': ['PROPOSED', 'RESERVED', 'PREPARED'],
            },
          },
        },
        'limit': 100,
      },
    );
    
    final orders = (response['orders'] as List?) ?? [];
    
    return orders.map((order) {
      final lineItems = (order['line_items'] as List?) ?? [];
      final totalMoney = order['total_money'] as Map<String, dynamic>?;
      final fulfillments = (order['fulfillments'] as List?) ?? [];
      
      String status = 'Pendiente';
      String type = 'Recoger';
      
      if (fulfillments.isNotEmpty) {
        final fulfillment = fulfillments[0] as Map<String, dynamic>;
        final state = fulfillment['state'] as String?;
        if (state == 'PREPARED') {
          status = 'Listo';
        } else if (state == 'RESERVED') {
          status = 'Procesando';
        }
        
        final typeValue = fulfillment['type'] as String?;
        if (typeValue == 'SHIPMENT') {
          type = 'Domicilio';
        }
      }
      
      return {
        'id': order['id'] ?? '',
        'orderNumber': (order['order_number'] ?? '').toString(),
        'customerName': order['recipient_name'] ?? 'Cliente',
        'itemCount': lineItems.length,
        'total': (totalMoney?['amount'] ?? 0) / 100.0,
        'status': status,
        'type': type,
        'createdAt': order['created_at'] ?? '',
      };
    }).toList();
  }
  
  // Obtener remesas (shipments/fulfillments)
  static Future<List<Map<String, dynamic>>> getShipments() async {
    await init();
    
    final response = await _makeRequest(
      '/v2/orders/search',
      'POST',
      {
        'location_ids': [_locationId],
        'query': {
          'filter': {
            'fulfillment_filter': {
              'fulfillment_types': ['SHIPMENT'],
            },
          },
        },
        'limit': 100,
      },
    );
    
    final orders = (response['orders'] as List?) ?? [];
    
    return orders.map((order) {
      final fulfillments = (order['fulfillments'] as List?) ?? [];
      final shipment = fulfillments.isNotEmpty ? fulfillments[0] as Map<String, dynamic> : null;
      
      return {
        'id': order['id'] ?? '',
        'orderNumber': (order['order_number'] ?? '').toString(),
        'customerName': order['recipient_name'] ?? 'Cliente',
        'status': shipment?['state'] ?? 'Pendiente',
        'trackingNumber': shipment?['shipment_details']?['carrier'] ?? '',
        'createdAt': order['created_at'] ?? '',
      };
    }).toList();
  }
}
